### PR TITLE
cmd,core,miner: EZDEV development+automining feature

### PR DESCRIPTION
**Pull request by whilei**
Creates flag --ezdev which
- uses 'ezdev' as --chain val, setting chaindata dir and other
related configuration and resource paths
- creates 10 accounts if keystore/ dir does not contain keyfiles
- sets respectable balance for each of these accounts in the genesis
block, then writing the genesis config file and other config files
to the chaindata dir
- creates AUTOMINING feature for miner, where addition of a transaction
to the txpool spawns a block win, which, with the automine config, is
set to bypass the nonce+mixhash header validations normally done for ethash
validations.
- turns on `--lightkdf` `--mine` and `--no-discover`

You can still use `--keystore` flag to specify the directory that should contain key files. You can still use `--sputnikvm`. You can still use the `--datadir` flag. 

The only real opinion here is that the `--ezdev` flag overrides the `--chain=ezdev` flag.

If you want to tinker with the config vals, take a look at the `chain.json` config file; you'll find that either in `<datadir>/ezdev/chain.json` or with `geth --ezdev dump-chain-config chain.json`.

### CODE REVIEW

File: miner/worker.go [hunk 1]
```diff
@@ -227,12 +231,36 @@ func (self *worker) update() {
...
+					// If many txs are fired at geth at once, they will be included in blocks more quickly than this event will be processed, causing ~200+ transactions per block, and yielding many empty blocks after the burst is complete as the channel drains. This conditional simply ensure that there are actually transactions to add a next block. If not, the event is essentially ignored.
+					pending, queued := self.eth.TxPool().Stats()
+					if pending+queued == 0 {
+						continue
+
+					}
+
+					self.currentMu.Lock()
+					ww := self.current
+					self.currentMu.Unlock()
```
Comment (tzdybal) [inline miner/worker.go]: Why it's not in single `Lock`? Why there is `Unlock` `noop` `Lock`?

File: miner/auto_agent.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	defer self.mu.Unlock()
+
+	close(self.quit)
+}
+
+func (self *AutoAgent) Start() {
+	self.mu.Lock()
+	defer self.mu.Unlock()
+
+	if !atomic.CompareAndSwapInt32(&self.isMining, 0, 1) {
```
Comment (tzdybal) [inline miner/auto_agent.go]: This code is protected by `self.mu` - is it necessary to use atomic operation in here?

File: miner/auto_agent.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+func (self *AutoAgent) SetReturnCh(ch chan<- *Result) { self.returnCh = ch }
+
+func (self *AutoAgent) Stop() {
+	self.mu.Lock()
+	defer self.mu.Unlock()
+
+	close(self.quit)
+}
+
+func (self *AutoAgent) Start() {
```
Comment (tzdybal) [inline miner/auto_agent.go]: `self` is considered bad name in go. `aa` or something more meaningful is preferred (I need a reference for this ;))

Comment (whilei) [reply miner/auto_agent.go]: Yea -- will fix for all the *agents.go. I just copy and pasted the remote_agent file which had that.

File: cmd/geth/cmd.go [hunk 4]
```diff
@@ -637,7 +638,7 @@ func rollback(ctx *cli.Context) error {
 func dumpChainConfig(ctx *cli.Context) error {
 
 	chainIdentity := mustMakeChainIdentity(ctx)
-	if !(core.ChainIdentitiesMain[chainIdentity] || core.ChainIdentitiesMorden[chainIdentity]) {
+	if !(core.ChainIdentitiesMain[chainIdentity] || core.ChainIdentitiesMorden[chainIdentity] || chainIdentity == "ezdev") {
```
[the remark this answers is not in the document]
Comment (whilei) [reply cmd/geth/cmd.go]: `ChainIdentitiesMain` is for mainnet aliases, as `ChainIdentitiesMorden` is for morden's (eg. `testnet`, etc).

Will addresss.

File: cmd/geth/ezdev.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	bal := "10000000000000000000000000000000"
+	if len(accman.Accounts()) == 0 {
+		glog.D(logger.Warn).Infoln("No existing EZDEV accounts found, creating 10")
+		password := ""
+		for i := 0; i < 10; i++ {
+			acc, err := accman.NewAccount(password)
+			if err != nil {
+				return err
+			}
+			a := strings.Replace(acc.Address.Hex(), "0x", "", -1)
```
[the remark this answers is not in the document]
Comment (whilei) [reply cmd/geth/ezdev.go]: Yea, good catch. This was just sloppy.

File: cmd/geth/flag.go [hunk 6]
```diff
@@ -198,6 +198,9 @@ func mustMakeChainIdentity(ctx *cli.Context) (identity string) {
 		}
 		// glog.V(logger.Debug).Infof("No existing file at --%v: '%v'. Using literal chain identity.", aliasableName(ChainIdentityFlag.Name, ctx), chainFlagVal)
 		identity = chainFlagVal
+
```
[the remark this answers is not in the document]
Comment (whilei) [reply cmd/geth/flag.go]: Good catch. Not an error. Just sloppy leftover line from debugging.

File: miner/auto_agent.go (hunk 2 above)
Comment (whilei) [reply miner/auto_agent.go]: This method again just copy and pasted from the `miner/agent.go@CpuAgent#update`. Not sure necessary, but IMO not worth messing with at this point.

File: miner/worker.go (hunk 1 above)
Comment (whilei) [reply miner/worker.go]: I had at one point thought that I might have seen a completely unquantifiable increase in performance. #nonoop

**Comment by lostinblue** [comment]
Mining is a big waste of energy, but I'm not against setting up a version of the client to simply setup a miner. If you need assistance with this please let me know Ill rebuild this client real quick to just mine.

**Pull request closed without merging**